### PR TITLE
feat(sqlalchemy): extend SQLAlchemy Plugin to include Alembic

### PR DESCRIPTION
**Pull request by cofin**
### Pull Request Checklist

[//]: # "By submitting this issue, you agree to:"
[//]: # "- follow Litestar's [Code of Conduct](https://github.com/litestar-org/.github/blob/main/CODE_OF_CONDUCT.md)"
[//]: # "- follow Litestar's [Contribution Guidelines](https://litestar.dev/community/contribution-guide)"

- [ ] New code has 100% test coverage
- [ ] (If applicable) The prose documentation has been updated to reflect the changes introduced by this PR
- [ ] (If applicable) The reference documentation has been updated to reflect the changes introduced by this PR

### Description
[//]: # "Please describe your pull request for new release changelog purposes"
[//]: # "Example: 'This pr adds the capability to use server-sent events, and documents it in the usage docs'"

-

### Close Issue(s)
[//]: # "Please add in issue numbers this pull request will close, if applicable"
[//]: # "Examples: Fixes #4321 or Closes #1234"

-

### CODE REVIEW

File: litestar/contrib/sqlalchemy/cli.py [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+    branch_label: str | None,
+    version_path: str | None,
+    rev_id: str | None,
+    no_prompt: bool,
+) -> None:
+    """Create a new database revision."""
+    if message is None and no_prompt:
+        message = "autogenerated"
+    elif message is None and not no_prompt:
+        message = prompt("Please enter a message describing this revision:")
```
Comment (provinzkraut) [inline litestar/contrib/sqlalchemy/cli.py:154]:
```suggestion
    if message is None:
        if no_prompt:
            message = "autogenerated"
        else:
            message = prompt("Please enter a message describing this revision")
```

File: litestar/contrib/sqlalchemy/cli.py [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+    is_flag=True,
+)
+def merge_revisions(
+    app: Litestar, revisions: str, message: str | None, branch_label: str | None, rev_id: str | None, no_prompt: bool
+) -> None:
+    """Create a new database revision."""
+    if message is None and no_prompt:
+        message = "autogenerated"
+    elif message is None and not no_prompt:
+        message = prompt("Please enter a message describing this revision")
```
Comment (provinzkraut) [inline litestar/contrib/sqlalchemy/cli.py:192]:
```suggestion
    if message is None:
        if no_prompt:
            message = "autogenerated"
        else:
            message = prompt("Please enter a message describing this revision")
```

File: litestar/plugins.py [hunk 3]
```diff
@@ -174,7 +174,7 @@ class PluginRegistry:
 
     def __init__(self, plugins: list[PluginProtocol]) -> None:
         self._plugins_by_type = {type(p): p for p in plugins}
-        self._plugins = frozenset(plugins)
+        self._plugins = set(plugins)
```
Comment (provinzkraut) [inline litestar/plugins.py:177]: Why this change?
The reason this was a frozenset is because we don't allow registering plugins dynamcially. If we want to, we need to take care of a bunch of stuff, e.g. regenerating the `openapi`, `init` and erialization` plugin tuples.

**Review by provinzkraut — CHANGES_REQUESTED** [review]
Looking really good!

File: litestar/plugins.py (hunk 3 above)
Comment (cofin) [reply litestar/plugins.py:177]: Sorry, this was not meant to be included here. This was part of my socketify debugging.

Comment (cofin) [reply litestar/plugins.py:177]:
```suggestion
        self._plugins = frozenset(plugins)
```

**Comment by cofin** [comment]
I still need to add tests, but I'm going to remove the draft status on this one.  I don't think there's anything additional features to add here.

File: litestar/contrib/sqlalchemy/alembic/commands.py [hunk 4]
```diff
@@ -71,10 +73,24 @@ class AlembicSpannerImpl(DefaultImpl):
...
 
 
+def _get_database_migration_plugin(app: Litestar) -> SQLAlchemyPlugin | SQLAlchemyInitPlugin:
+    """Retrieve a database migration plugin from the Litestar application's plugins.
+
+    This function attempts to find and return either the SQLAlchemyPlugin or SQLAlchemyInitPlugin.
+    If neither plugin is found, it raises an ImproperlyConfiguredException.
+    """
+    for type_ in SQLAlchemyPlugin, SQLAlchemyInitPlugin:
+        with suppress(KeyError):
```
Comment (provinzkraut) [inline litestar/contrib/sqlalchemy/alembic/commands.py:83]:
```suggestion
        if type_ in app.plugins:
```

Comment (Alc-Alc) [reply litestar/contrib/sqlalchemy/alembic/commands.py:83]: > Suggested change
>         with suppress(KeyError):
>         if type_ in app.plugins:

I think this does not exactly do what I intend to? It always seems to trigger `ImproperlyConfiguredException`

**Comment by Alc-Alc** [comment]
In commit [cffc81b](https://github.com/litestar-org/litestar/pull/2164/commits/cffc81b1386043f45adefe2fbe285851d7ffb8da), the addition of `_namespace_metadata=op._namespace_metadata` to `order_columns` addresses the `AttributeError: 'NoneType' object has no attribute 'schema'` issue. This workaround will be necessary until the resolution of [sqlalchemy/alembic issue #1193](https://github.com/sqlalchemy/alembic/issues/1193) to ensure that custom column sorting within CREATE TABLE works.

**Comment by Faolain** [comment]
Is there any reason why this was closed and the branch deleted? It appears a significant amount of work was put into this and it was just about done unless I'm misunderstanding something here.

**Comment by Goldziher** [comment]
@Faolain - cofin stepped down as maintainer. You can see the discussion in our discord server. 

I restored the branch for now.
